pkg/jobs/update.go

```diff
@@ -379,8 +379,8 @@ func getSelectStmtForJobUpdate(
 	const (
 		selectWithoutSession = `
 WITH
-	latestpayload AS (SELECT job_id, value FROM system.job_info AS payload WHERE info_key = 'legacy_payload'::BYTES),
-	latestprogress AS (SELECT job_id, value FROM system.job_info AS progress WHERE info_key = 'legacy_progress'::BYTES)
+	latestpayload AS (SELECT job_id, value FROM system.job_info AS payload WHERE info_key = 'legacy_payload'::BYTES AND job_id = $1 ORDER BY written DESC LIMIT 1),
```

Comment:
This is strange. `TestBackupTenantImportingTable` doesn't like the LIMIT 1 clause at the end of this query. After the restore in that test when we restart the tenant the test immediately fails with:

```
E230328 23:51:18.245401 6288 2@rpc/context.go:2388  [T10,nsql2,rnode=1,raddr=127.0.0.1:62608,class=default,rpc] 335  unable to connect (is the peer up and reachable?): initial connection heartbeat failed: grpc: co
nnection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:62608: connect: connection refused" [code 14/Unavailable]
I230328 23:51:18.245434 6288 rpc/context.go:2302  [T10,nsql2,rnode=1,raddr=127.0.0.1:62608,class=default,rpc] 336  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection er
ror: desc = "transport: error while dialing: dial tcp 127.0.0.1:62608: connect: connection refused" [code 14/Unavailable]
I230328 23:51:18.245467 6858 1@circuitbreaker/circuitbreaker.go:322  [T10,nsql2] 337  circuitbreaker: rpc 127.0.0.1:62620 [n1] tripped: failed to connect to n1 at 127.0.0.1:62608: initial connection heartbeat fail
ed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:62608: connect: connection refused" [code 14/Unavailable]
I230328 23:51:18.245503 6858 1@circuitbreaker/circuitbreaker.go:447  [T10,nsql2] 338  circuitbreaker: rpc 127.0.0.1:62620 [n1] event: BreakerTripped
W230328 23:51:18.245473 6514 2@rpc/nodedialer/nodedialer.go:196  [T10,nsql2,client=127.0.0.1:62626,user=root,intExec=select-job] 339  unable to connect to n1: failed to connect to n1 at 127.0.0.1:62608: initial co
nnection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:62608: connect: connection refused" [code 14/Unavailable]
W230328 23:51:18.245581 6858 sql/colflow/colrpc/outbox.go:183  [T10,nsql2,client=127.0.0.1:62626,user=root,intExec=select-job,f39c33432,distsql.stmt=WITH latestpayload AS (SELECT job_id, value FROM system.job_info
 AS payload WHERE (info_key = '_'::BYTES) AND (job_id = $1) ORDER BY written DESC LIMIT _), latestprogress AS (SELECT job_id, value FROM system.job_info AS progress WHERE (info_key = '_'::BYTES) AND (job_id = $1)
ORDER BY written DESC LIMIT _) SELECT status, payload.value AS payload, progress.value AS progress, COALESCE(last_run, created), COALESCE(num_runs, _) FROM system.jobs AS j INNER JOIN latestpayload AS payload ON j
.id = payl,distsql.gateway=2,distsql.appname=$ internal-select-job,distsql.txn=31defc85-a5e6-4cdc-97ca-ed6431fd2104,streamID=6] 340  Outbox Dial connection error, distributed query will fail: failed to connect to
n1 at 127.0.0.1:62608: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:62608: connect: connection refused" [code 14/Unavailable]
```

Dropping the LIMIT 1 "fixes" it 🤔

Reply:
I would be lying if I said I knew what was exactly going on but pushing the CANCEL into the succeeds soon allows the test to succeed. Why a LIMIT 1 clause in a CTE would so reliably make the CANCEL fail is still a mystery. This test does stop the tenant server, and restart it post-restore so it's possible this is related to the various slow quiesce issues we're seeing elsewhere.

Reply:
I'm actually going to go out on a limb and say whatever Yahor finds https://github.com/cockroachdb/cockroach/issues/99753 is going to be related. Pattern matching `Outbox Dial connection error, distributed query will fail: dial: context canceled` to that issue